rfcs/20230110-kernel-reuse/README.md

```diff
@@ -0,0 +1,286 @@
+
+#### Cons:
+Requires a finite number of generators, no guarantee generators align with
+customer workloads.
+
+Users do not control the registry size
+
+Requires the generator to be GPU agnostic (i.e. should work for 2 or more GPUs
+in a system if they have the same architecture) to limit the registry size.
+
+Implementation needs to limit maximum memory usage to a reasonable level.
+
+#### Implementation
+This feature will be completely managed by oneDNN developers. As such, there are
+no API changes. The only guarantee is that the registry is generated at runtime
+and will not use much memory until users creates primitives. It is expected the
+implementation will be a hashmap between a kernel configuration and a kernel
+generator. Registries will have a compile time defined maximum size to avoid
+excessive memory usage. As a registries size is dependent on implementation,
+separate registries will exist per implementation. As the requirements on this
+implementations are similar to Distribution Option 1,this can be used to enable
+build time compilation.
+
+### Reuse Option 4: Global Kernel Generator Cache
+This options is a version of the global primitive cache, but for gpu kernel
+generators. Kernel generators are expected to be performant. For OpenCL C
+kernels, the generator is expected to be a program binary.
+
+#### Pros:
+Requires no user intervention.
+
+Implementation for global primitive cache is similar.
+
+Users can control memory usage
+
+Generators can be specialized for each GPU in a system, potentially enabling
+faster performance.
+
+#### Cons:
+Most complicated implementation (except potentially Option 2) among all the
+options.
+
+#### Implementation
+The API change will be the same as in the [global primitive
+cache](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20200207-global-primitive-cache),
+except the "primitive" name will be replaced with "kernel". Kernels which use
+this cache will need to opt in to its uses. Kernels need to be designed for
+reuse for the cache to be effective.
+
+### Recommendation
+Short Answer: Start with Reuse Option 3 with a transition to a combination of
+Reuse Option 2 and Reuse Option 4 on an as need basis.
+
+There are significant benefits to modifying our implementations to generate a
+limited and finite number of kernel binaries. Among the benefits are:
+* Enables Build Time Compilation
+* Enables Runtime Parameters
+* Reduces Testing Complexity
+
+While it is impossible to achieve this in general, for sufficiently large
+problems, there tend to only be a few important compiled configurations. Because
+of these advantages, I recommend we start curating a set of kernels which
+address most sufficiently large problems.
+
+To reuse, I recommend starting with Reuse Option 3, the fixed size function
+registry. The motivating reason is the kernels generated for Reuse Option 3 also
+satisfy the requirements for Distribution Option 1, enabling us to address
```

Comment:
@rjoursler quick question, how option 3 address first compilation issue? 
You mention in option3 that generation happens on runtime, my understanding for the flow is: primitive creation -> get kernel generator if cached, create if not cached -> generate kernel, is it correct? 
If yes, for first compilation, we still see the overhead because kernel generator is not there, right? Anything I missed? Thanks!

Reply:
You are correct. In fact, there will be a first compilation issue with all the reuse options. The only way to completely remove first compilation for OpenCL kernels is to implement Distribution Option 1, where we move compilation to build time and distribute the binaries within the library.  This is currently infeasible as, at the very least, we would need to enable effective kernel reuse and simultaneously keep library a reasonable size.

Instead, the reuse options decrease the number of first compilations. The current implementation performs compilations on a per-primitive basis. For Options 3 and 4 the target is to compile on a per-kernel basis. This will allow us to reduce the number of first compilations to the point it is irrelevant for most use cases.

Reply:
Thanks for explanation.